### PR TITLE
[Internal Balances] create useInternalBalancesTransaction

### PR DESCRIPTION
**Pull request by luizakp**
This PR has the goal to reduce duplicated code between useTransactionMetadata and useInternalBalancesTransactions

### CODE REVIEW

File: packages/next-app/src/app/internalmanager/page.tsx [hunk 1]
```diff
@@ -125,138 +47,82 @@ export default function Page() {
             <Table.Body>
               {tokensWithBalance.map((token) => (
                 <TableRow
-                  key={token.tokenInfo.address}
-                  token={token}
+                  key={token.token}
+                  //TODO change to useInternalBalances when subgraph is updated
+                  //https://linear.app/bleu-llc/issue/BAL-272/fix-internal-balance-token-attribute-on-subgraph
+                  token={token as any}
```
Comment (ribeirojose) [inline packages/next-app/src/app/internalmanager/page.tsx:53]: 👁️

File: packages/next-app/src/app/internalmanager/page.tsx [hunk 2]
```diff
@@ -125,138 +47,82 @@ export default function Page() {
...
-        ...prev,
-        status: TransactionStatus.WRITE_ERROR,
-      }));
-    },
+  const {
+    transactionUrl,
+    handleWithdraw,
+    isNotifierOpen,
+    setIsNotifierOpen,
+    notification,
```
Comment (ribeirojose) [inline packages/next-app/src/app/internalmanager/page.tsx:79]: Shouldn't these things be handled by the hook itself?

File: packages/next-app/src/hooks/useTransaction.ts [hunk 3]
```diff
@@ -75,13 +80,40 @@ const NOTIFICATION_MAP = {
...
+export const networkUrls = {
+  [NetworkChainId.MAINNET]: "https://etherscan.io/tx/",
+  [NetworkChainId.GOERLI]: "https://goerli.etherscan.io/tx/",
+  [NetworkChainId.POLYGON]: "https://polygonscan.com/tx/",
+  [NetworkChainId.ARBITRUM]: "https://arbiscan.io/tx/",
 };
 
+function getNetworkUrl(chainId: NetworkChainId) {
+  return networkUrls[chainId as keyof typeof networkUrls];
+}
```
Comment (ribeirojose) [inline packages/next-app/src/hooks/useTransaction.ts:115]: this should probably live in the `shared` package (if it doesn't already)

File: packages/next-app/src/hooks/useTransaction.ts [hunk 4]
```diff
@@ -190,3 +222,109 @@ export function useMetadataTransaction({
...
+    setNotification(
+      NOTIFICATION_MAP_INTERNAL_BALANCES[TransactionStatus.WAITING_APPROVAL]
+    );
+  }
+
+  //trigger the actual transaction
+  useEffect(() => {
+    if (!operationKind) return;
+    write?.();
+  }, [operationKind]);
```
Comment (ribeirojose) [inline packages/next-app/src/hooks/useTransaction.ts:268]: Shouldn't this be inside `handleWithdraw`? Why trigger `write` magically here?

File: packages/next-app/src/hooks/useTransaction.ts [hunk 5]
```diff
@@ -190,3 +222,109 @@ export function useMetadataTransaction({
...
+
+  //handle transaction status
```
Comment (ribeirojose) [inline packages/next-app/src/hooks/useTransaction.ts:270]: 👁️

File: packages/next-app/src/hooks/useTransaction.ts [hunk 6]
```diff
@@ -190,3 +222,109 @@ export function useMetadataTransaction({
...
+      case UserBalanceOpKind.WITHDRAW_INTERNAL: {
+        handleTransactionStatus();
+      }
+      default:
+        return;
+    }
+  }, [data]);
+
+  //check if transaction is confirmed
+  useWaitForTransaction({
```
Comment (devjoaov) [inline packages/next-app/src/hooks/useTransaction.ts:292]: 👍

File: packages/next-app/src/hooks/useTransaction.ts [hunk 7]
```diff
@@ -190,3 +222,109 @@ export function useMetadataTransaction({
...
+    if (!operationKind) return;
+    write?.();
+  }, [operationKind]);
+
+  //handle transaction status
+  useEffect(() => {
+    if (!data) return;
+    const { hash } = data;
+    async function handleTransactionStatus() {
+      if (!hash || !chain) return;
```
Comment (devjoaov) [inline packages/next-app/src/hooks/useTransaction.ts:275]: i would move this to the `switch case` block

**Review by devjoaov — APPROVED** [review]

File: packages/next-app/src/hooks/useTransaction.ts [hunk 8]
```diff
@@ -190,3 +222,109 @@ export function useMetadataTransaction({
...
+  const [notification, setNotification] = useState<Notification | null>(null);
+  const [transactionUrl, setTransactionUrl] = useState<string>();
+  const [operationKind, setOperationKind] = useState<UserBalanceOpKind>();
+
+  //Prepare data for transaction
+  const userBalanceOp = {
+    kind: operationKind as number,
+    asset: token.token as `0x${string}`,
+    amount: parseFixed(token.balance, 18),
+    sender: userAddress as `0x${string}`,
```
[the remark this answers is not in the document]
Comment (luizakp) [reply packages/next-app/src/hooks/useTransaction.ts:246]: from the `InternalBalanceQuery` genereted from graphql the address comes as `string` but the contract expects `0x${string}`

File: packages/next-app/src/hooks/useTransaction.ts [hunk 9]
```diff
@@ -190,3 +222,109 @@ export function useMetadataTransaction({
...
+  useEffect(() => {
+    if (!operationKind) return;
+    write?.();
+  }, [operationKind]);
+
+  //handle transaction status
+  useEffect(() => {
+    if (!data) return;
+    const { hash } = data;
+    async function handleTransactionStatus() {
```
[the remark this answers is not in the document]
Comment (luizakp) [reply packages/next-app/src/hooks/useTransaction.ts:274]: removed it hehehhe it used to have a `wait` that was a Promise. Thanks for pointing that out 👍

File: packages/next-app/src/hooks/useTransaction.ts (hunk 4 above)
Comment (luizakp) [reply packages/next-app/src/hooks/useTransaction.ts:268]: Will solve this [here](https://linear.app/bleu-llc/issue/BAL-249/implement-internal-balance-withdraw-flow)

File: packages/next-app/src/hooks/useTransaction.ts (hunk 5 above)
Comment (luizakp) [reply packages/next-app/src/hooks/useTransaction.ts:270]: Will solve this [here](https://linear.app/bleu-llc/issue/BAL-249/implement-internal-balance-withdraw-flow)

**Pull request merged**